---
title: "R_assignment"
output: html_document
---

# PART 1

# 0. Load tidyverse
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
library(tidyr)
```

# 1. Import fang_et_al_genotypes and snp_position and analyze contents.
```{r}
genotypes <- read_tsv("./fang_et_al_genotypes.txt") 

snp_positions <- read_tsv("./snp_position.txt")

#file size
cat("\nFile Size\n")
paste0("fang_et_al_genotypes.txt is ",toString(object.size(genotypes))," bytes.")
paste0("snp_position.txt is ",toString(object.size(snp_positions))," bytes.")

cat("\nNumber of Columns\n")
paste0("fang_et_al_genotypes.txt has ", toString(ncol(genotypes))," columns.")
paste0("snp_position.txt has ", toString(ncol(snp_positions))," columns.")

cat("\nNumber of Rows\n")
paste0("fang_et_al_genotypes.txt has ", toString(nrow(genotypes))," rows.")
paste0("snp_position.txt has ", toString(nrow(snp_positions))," rows.")

```

# 2. Take columns 1,3,4 from snp_position and save as new variable.
```{r}
snp_134 <- snp_positions[c(1,3,4)]
rm(snp_positions)
```

# 3. Separate out data from Teosinte and Maize, respectively.
```{r}
#Teosinte
Teo_genotypes <- filter(genotypes, `Group` %in% c("ZMPBA","ZMPIL","ZMPJA"))

#Maize
Maize_genotypes <- filter(genotypes, `Group` %in% c("ZMMIL","ZMMLR"," ZMMMR"))
rm(genotypes)

```

# 4. Transpose gene files created above, rename 'sample_ID' to 'SNP_ID'
```{r}
#Teosinte
t_Teosinte <- as.data.frame(t(Teo_genotypes))

#change SNP row names to its own column
t_Teosinte <- rownames_to_column(t_Teosinte,"SNP_ID")
print(t_Teosinte[1:3,])
#Remove unnecessary rows
t_Teosinte <- t_Teosinte[-(1:3),]


#Maize
t_Maize <- as.data.frame(t(Maize_genotypes))

#change SNP row names to its own column
t_Maize <- rownames_to_column(t_Maize,"SNP_ID")
print(t_Maize[1:3,])
#Remove unnecessary rows
t_Maize <- t_Maize[-(1:3),]
print(t_Maize[1:3,])
```

# 5. Arrange all three files by SNP name.
```{r}
#SNP positions
snp_sorted <- arrange(snp_134, `SNP_ID`)
rm(snp_134)

#Teosinte
Teo_sorted <- arrange(t_Teosinte,`SNP_ID`)
rm(t_Teosinte)
#Maize
Maize_sorted <- arrange(t_Maize,`SNP_ID`)
rm(t_Maize)
```

# 6. Merge teosinte and maize files each with SNP file. 
```{r}
#Teosinte
Teo_merge_q <- merge(snp_sorted,Teo_sorted, all=TRUE)
Teo_merge_q[1:5]

rm(Teo_sorted)


#Maize
Maize_merge_q <- merge(snp_sorted,Maize_sorted, all=TRUE)
Maize_merge_q[1:5]

rm(snp_sorted)
rm(Maize_sorted)
```

# 7. Separate merged files' chromosomes into folders and arrange.
```{r}
#Teosinte
for (i in 1:10){
  filename = paste0('Teosinte_increasing_',i,'.txt')
  #filter by chromosome number
   x = filter(Teo_merge_q, `Chromosome` == i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
   x_arr = arrange(x,`Position`)
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}
#Maize
for (i in 1:10){
  filename = paste0('Maize_increasing_',i,'.txt')
  #filter by chromosome number
  x = filter(Maize_merge_q, `Chromosome`== i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
  x_arr = arrange(x,`Position`)
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}
read_tsv("Maize_increasing_1.txt")

```

# 8. Separate merged files' chromosomes into folders and arrange. Create new file for each where the sign for missing data is changed from ?/? to -/-
```{r}
#Teosinte
for (i in 1:10){
  filename = paste0('Teosinte_decreasing_',i,'.txt')
  #filter by chromosome number
   x = filter(Teo_merge_q, `Chromosome` == i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
   x_arr = arrange(x,desc(`Position`))
   x_arr <- as.data.frame(lapply(x_arr, gsub, pattern="\\?", replacement="-"))
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}

read_tsv("Teosinte_decreasing_1.txt")

#Maize
for (i in 1:10){
  filename = paste0('Maize_decreasing_',i,'.txt')
  #filter by chromosome number
  x = filter(Maize_merge_q, `Chromosome`== i)
  #convert position values from characters to numbers
  x$Position <- as.numeric(x$Position)
  #order by position
  x_arr = arrange(x,desc(`Position`))
  x_arr <- as.data.frame(lapply(x_arr, gsub, pattern="\\?", replacement="-"))
  write.table(x_arr, file = filename,sep = '\t',quote=FALSE, row.names=FALSE)
}

```



# PART 2

# 1. SNPs per Chromosome
# Plot total number of SNPs in our dataset on each chromosome
```{r}
#Teosinte
snp_count_teo <- ggplot(data = Teo_merge_q) +
  geom_bar(mapping = aes(x = Chromosome, fill= Chromosome))
snp_count_teo <- snp_count_teo + ylab("SNP Count") + xlab("Teosinte Chromosome") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 
#+ scale_x_discrete(limits=c("1","2","3","4","5","6","7","8","9","10"))
print(snp_count_teo)


#Maize
snp_count_maize <- ggplot(data = Maize_merge_q) +
  geom_bar(mapping = aes(x = Chromosome,fill= Chromosome ))
snp_count_maize <- snp_count_maize + ylab("SNP Count") + xlab("Maize Chromosome") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 
#+ scale_x_discrete(limits=c("1","2","3","4","5","6","7","8","9","10"))
print(snp_count_maize)

```


# Plot distribution of SNPs on Chromosomes
```{r}
#Teosinte
Teo_merge_q$Position <- as.numeric(Teo_merge_q$Position)
SNP_dens_Teo <- ggplot(data = Teo_merge_q) + geom_density(mapping = aes(x=Position, fill=Chromosome),alpha =0.2) + facet_wrap( ~ Chromosome,ncol=3)
print(SNP_dens_Teo)

#Maize
Maize_merge_q$Position <- as.numeric(Maize_merge_q$Position)
SNP_dens_Maize <- ggplot(data = Maize_merge_q) + geom_density(mapping = aes(x=Position, fill=Chromosome),alpha =0.2) + facet_wrap( ~ Chromosome,ncol=3)
print(SNP_dens_Maize)
```

# 2. Heterozygosity
# Separate genotypes data into Teosinte & Maize data, pivot datasets.
```{r}
genotypes <- read_tsv("./fang_et_al_genotypes.txt") 

Teo_genotypes <- filter(genotypes, `Group` %in% c("ZMPBA","ZMPIL","ZMPJA"))

Maize_genotypes <- filter(genotypes, `Group` %in% c("ZMMIL","ZMMLR"," ZMMMR"))
rm(genotypes)

#pivot both sets
Teo_pivot <- pivot_longer(Teo_genotypes, -Sample_ID:-Group, names_to = "SNP_ID", values_to = "Sequence")

Maize_pivot <- pivot_longer(Maize_genotypes, -Sample_ID:-Group, names_to = "SNP_ID", values_to = "Sequence")

rm(Teo_genotypes)
rm(Maize_genotypes)

Maize_pivot

```


# Convert pivoted data to a matrix so its SNPs can be replaced with "Homo", "Hetero" or "Unknown"
```{r}
#Teosinte
Teo_p_matrix = as.matrix(Teo_pivot)

Teo_pivot_homo <- replace(Teo_p_matrix, Teo_p_matrix %in% c("T/T", "C/C", "A/A", "G/G"), "Homo")
Teo_pivot_hetero <- replace(Teo_pivot_homo, Teo_pivot_homo %in% c("A/T", "A/C", "A/G", "T/A", "T/C", "T/G", "C/A", "C/G", "C/T", "G/A", "G/C", "G/T"), "Hetero")
Teo_heterozygosity <- as.data.frame(replace(Teo_pivot_hetero, Teo_pivot_hetero %in% c("?/?", "?/A", "A/?", "T/?", "?/T", "C/?", "?/C", "G/?", "?/G"), "Unknown"))

rm(Teo_p_matrix)
rm(Teo_pivot_hetero)
rm(Teo_pivot_homo)

#Maize
Maize_p_matrix = as.matrix(Maize_pivot)

M_pivot_homo <- replace(Maize_p_matrix, Maize_p_matrix %in% c("T/T", "C/C", "A/A", "G/G"), "Homo")
M_pivot_hetero <- replace(M_pivot_homo, M_pivot_homo %in% c("A/T", "A/C", "A/G", "T/A", "T/C", "T/G", "C/A", "C/G", "C/T", "G/A", "G/C", "G/T"), "Hetero")
Maize_heterozygosity <- as.data.frame(replace(M_pivot_hetero, M_pivot_hetero %in% c("?/?", "?/A", "A/?", "T/?", "?/T", "C/?", "?/C", "G/?", "?/G"), "Unknown"))

rm(Maize_p_matrix)
rm(M_pivot_hetero)
rm(M_pivot_homo)

```


# Graph modified heterozygosity data by Group and Sample ID using ggplot
```{r}
#Teosinte

#Graph heterozygosity vs Group
ggplot(data = Teo_heterozygosity) + geom_bar(mapping = aes(x = Teo_heterozygosity$Group, fill = Teo_heterozygosity$Sequence), position = "fill")
#Graph heterozygosity vs Sample_ID
ggplot(data = Teo_heterozygosity) + geom_bar(mapping = aes(x = Teo_heterozygosity$Sample_ID, fill = Teo_heterozygosity$Sequence), position = "fill")



#Maize

#Graph heterozygosity vs Group
ggplot(data = Maize_heterozygosity) + geom_bar(mapping = aes(x = Maize_heterozygosity$Group, fill = Maize_heterozygosity$Sequence), position = "fill")
#Graph heterozygosity vs Sample_ID
ggplot(data = Maize_heterozygosity) + geom_bar(mapping = aes(x = Maize_heterozygosity$Sample_ID, fill = Maize_heterozygosity$Sequence), position = "fill")
```



# My plot
# Amount of Heterozygosity by Chromosome
```{r}
#pivot both sets
Teo_myplot <- pivot_longer(Teo_merge_q, -SNP_ID:-Position, names_to = "names", values_to = "Sequence")

Maize_myplot <- pivot_longer(Maize_merge_q, -SNP_ID:-Position, names_to = "names", values_to = "Sequence")

# Convert pivoted data to a matrix so its SNPs can be replaced with "Homo", "Hetero" or "Unknown"

#Teosinte
Teo_p_matrix = as.matrix(Teo_myplot)
rm(Teo_myplot)

Teo_pivot_homo <- replace(Teo_p_matrix, Teo_p_matrix %in% c("T/T", "C/C", "A/A", "G/G"), "Homo")
Teo_pivot_hetero <- replace(Teo_pivot_homo, Teo_pivot_homo %in% c("A/T", "A/C", "A/G", "T/A", "T/C", "T/G", "C/A", "C/G", "C/T", "G/A", "G/C", "G/T"), "Hetero")
Teo_chromosome_heterozygosity <- as.data.frame(replace(Teo_pivot_hetero, Teo_pivot_hetero %in% c("?/?", "?/A", "A/?", "T/?", "?/T", "C/?", "?/C", "G/?", "?/G"), "Unknown"))

rm(Teo_p_matrix)
rm(Teo_pivot_hetero)
rm(Teo_pivot_homo)

#Maize
Maize_p_matrix = as.matrix(Maize_myplot)
rm(Maize_myplot)

M_pivot_homo <- replace(Maize_p_matrix, Maize_p_matrix %in% c("T/T", "C/C", "A/A", "G/G"), "Homo")
M_pivot_hetero <- replace(M_pivot_homo, M_pivot_homo %in% c("A/T", "A/C", "A/G", "T/A", "T/C", "T/G", "C/A", "C/G", "C/T", "G/A", "G/C", "G/T"), "Hetero")
Maize_chromosome_heterozygosity <- as.data.frame(replace(M_pivot_hetero, M_pivot_hetero %in% c("?/?", "?/A", "A/?", "T/?", "?/T", "C/?", "?/C", "G/?", "?/G"), "Unknown"))

rm(Maize_p_matrix)
rm(M_pivot_hetero)
rm(M_pivot_homo)


#Graph Heterozygosity vs Chromsome

#Teosinte
ggplot(data = Teo_chromosome_heterozygosity) + geom_bar(mapping = aes(x = Chromosome, fill = Teo_chromosome_heterozygosity$Sequence), position = "fill") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 

#Maize
ggplot(data = Maize_chromosome_heterozygosity) + geom_bar(mapping = aes(x = Chromosome, fill = Maize_chromosome_heterozygosity$Sequence), position = "fill") + scale_x_discrete(labels=c("1","10","2","3","4","5","6","7","8","9","Mult.","Unk")) 

```

